Common/include/option_structure.inl

```diff
@@ -26,48 +26,50 @@
  * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
  */
 
-template <class Tenum>
-class COptionEnum : public COptionBase {
+template <class Tenum, class TField>
+class COptionEnum final : public COptionBase {
 
-  map<string, Tenum> m;
-  unsigned short & field; // Reference to the feildname
-  Tenum def; // Default value
-  string name; // identifier for the option
+  const map<string, Tenum>& m;
+  TField& field; // Reference to the fieldname
+  const Tenum def; // Default value
+  const string name; // identifier for the option
 
 public:
-  COptionEnum(string option_field_name, const map<string, Tenum> m, unsigned short & option_field, Tenum default_value) : field(option_field) {
-    this->m = m;
-    this->def = default_value;
-    this->name = option_field_name;
+  COptionEnum() = delete;
+
+  COptionEnum(string option_field_name, const map<string, Tenum>& m_, TField& option_field, Tenum default_value) :
+    m(m_),
+    field(option_field),
+    def(default_value),
+    name(std::move(option_field_name)) {
   }
 
-  ~COptionEnum() override {};
+  ~COptionEnum() = default;
+
   string SetValue(vector<string> option_value) override {
     COptionBase::SetValue(option_value);
     // Check if there is more than one string
-    string out = optionCheckMultipleValues(option_value, "enum", this->name);
+    string out = optionCheckMultipleValues(option_value, "enum", name);
     if (out.compare("") != 0) {
       return out;
     }
 
     // Check to see if the enum value is in the map
-    if (this->m.find(option_value[0]) == m.end()) {
-      string str;
-      str.append(this->name);
+    auto it = m.find(option_value[0]);
+
+    if (it == m.cend()) {
+      string str = name;
       str.append(": invalid option value ");
       str.append(option_value[0]);
       str.append(". Check current SU2 options in config_template.cfg.");
       return str;
     }
     // If it is there, set the option value
-    Tenum val = this->m[option_value[0]];
-    this->field = val;
+    field = static_cast<TField>(it->second);
     return "";
   }
 
-  void SetDefault() override {
-    this->field = this->def;
-  }
+  void SetDefault() override { field = static_cast<TField>(def); }
```

Comment:
I added the static cast but this was not smart :facepalm: (it would make enum class compatible with unsigned short...)

Reply:
sorted